site/content/blog/2021-04-23-optimizing-go-compiler-1.md

```diff
@@ -0,0 +1,227 @@
+
+```go
+var c uint8 = a + b
+c = c - b
+```
+
+This can be represented in SSA as something like:
+
+```
+v4 = Add8 <uint8> v2 v3
+v5 = Sub8 <uint8> v4 v3
+```
+
+In this case, `v2` is a _value_ that represents the variable `a`, while `v3` is a value representing the variable `b`. We store its addition---represented by the operator `Add8` with type `uint8`--- into the new value `v4`.
+
+Then, in the second line of the original code, we reassign `c`. But remember: every value in SSA is assigned **only once**. Thus, in SSA, we generate a new value `v5`, that will contain `c` (`v4`) minus `b` (`v3`).
+
+The best way to learn about SSA is generating it ourselves. Thankfully, this is one shell command away. Whenever you are building or running a piece of Go code, tell the compiler that you want to read the generated SSA by setting the env variable `GOSSAFUNC` to the name of the function you want to investigate. It looks something like:
+
+```sh
+GOSSAFUNC=myFunction go build myfile.go
+```
+
+This outputs a `ssa.html` file, in the same directory where you ran the command, that you can open with your browser. It looks like this:
+
+![Screenshot of the SSA page](/blog/2021-04-23-optimizing-go-compiler-1/ssa.png)
+
+That's a whole lot of information. I agree. But we can focus on what's interesting for us. The page shows the output of every phase of the compilation, once SSA is generated, in collapsible columns. To make things easier, the page is interactive: try clicking on a value and see how the uses of that value are highlighted in the same color. This does not only happen within the same column: if you look at the previous or next columns, you can also quickly see how that value is being modified by the different phases of the compilation (or if it is left untouched). You can do the same thing with blocks (the labels starting with `b`), which are isolated series of values in the control flow graph of the function.
+
+## Rewrite rules
+
+One of the benefits of using SSA is that it is easy to match common patterns and rewrite them with optimized versions. For example, let's say that we have the following Go code:
+
+```go
+var five = 5
+var a = five * 16
+```
+
+We can compute the value of `a` by literally making the multiplication `5 times 16`. But it is known that multiplication instructions are quite CPU-expensive. What if we could optimize this line of code by avoiding the multiplication?
+
+It turns out we can, because `16` is a power of `2`. And multiplying by `2^n` is the same as shifting the number to the left `n` places. This is easy to see in binary, where 5 is:
+
+```
+00000101
+```
+
+If we multiply 5 by 2, we get 10, which in binary is:
+
+```
+00001010
+```
+
+And again by 2, we get 20:
+
+```
+00010100
+```
+
+See how we are shifting everything to the left each time we multiply by 2? Then, multiplying by 16 is the same as shifting to the left 4 times, as 16 is `2^4`:
+
+```
+00000101 // 5
+01010000 // 5*(2^4) = 80
+```
```

Comment:
I feel like this level of detail is unnecessary. We can just skip this part with just a one-line explanation that doubling is the same as a single left shift.

Reply:
I agree. I erred on the side of over-explaining, and this is a good example. Changed, thanks!